File: banchus-app/src/main/java/pe/nanamochi/banchus/service/BeatmapsetService.java
package pe.nanamochi.banchus.service;

import java.util.Optional;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import pe.nanamochi.banchus.database.entity.Beatmapset;
import pe.nanamochi.banchus.database.repository.BeatmapsetRepository;
import pe.nanamochi.banchus.domain.dto.OsuApiBeatmap;
import pe.nanamochi.banchus.mapper.BeatmapsetMapper;

@Service
@RequiredArgsConstructor
public class BeatmapsetService {
  private final BeatmapsetRepository beatmapsetRepository;
  private final BeatmapsetMapper beatmapsetMapper;

  public Beatmapset save(Beatmapset beatmapset) {
    return beatmapsetRepository.save(beatmapset);
  }

  public Beatmapset save(OsuApiBeatmap apiBeatmap) {
    Beatmapset beatmapset = beatmapsetMapper.fromApi(apiBeatmap);
    return save(beatmapset);
  }

  public Beatmapset update(Beatmapset beatmapset) {
    if (!beatmapsetRepository.existsById(beatmapset.getId())) {
      throw new IllegalArgumentException("Beatmapset not found: " + beatmapset.getId());
    }
    return beatmapsetRepository.save(beatmapset);
  }

  public Optional<Beatmapset> findById(int id) {
    return beatmapsetRepository.findById(id);
  }
}

File: banchus-app/src/main/kotlin/pe/nanamochi/banchus/service/BeatmapsetService.kt
package pe.nanamochi.banchus.service

import com.github.michaelbull.result.Err
import com.github.michaelbull.result.Result
import com.github.michaelbull.result.toResultOr
import org.springframework.stereotype.Service
import pe.nanamochi.banchus.database.entity.Beatmapset
import pe.nanamochi.banchus.database.repository.BeatmapsetRepository
import pe.nanamochi.banchus.domain.errors.BeatmapsetNotFound
import pe.nanamochi.banchus.domain.errors.DomainMessage
import pe.nanamochi.banchus.dto.external.OsuApiBeatmap
import pe.nanamochi.banchus.mapper.BeatmapsetMapper
import pe.nanamochi.banchus.util.runDatabaseCatching

@Service
class BeatmapsetService(
    private val beatmapsetRepository: BeatmapsetRepository,
    private val beatmapsetMapper: BeatmapsetMapper,
) {
    fun findById(id: Int): Result<Beatmapset, BeatmapsetNotFound> =
        beatmapsetRepository.findBeatmapsetById(id).toResultOr { BeatmapsetNotFound }

    fun create(beatmapset: Beatmapset): Result<Beatmapset, DomainMessage> = runDatabaseCatching {
        beatmapsetRepository.save(beatmapset)
    }

    fun create(apiBeatmap: OsuApiBeatmap): Result<Beatmapset, DomainMessage> =
        create(beatmapsetMapper.fromApi(apiBeatmap))

    fun update(beatmapset: Beatmapset): Result<Beatmapset, DomainMessage> =
        if (beatmapsetRepository.existsById(beatmapset.id)) {
            runDatabaseCatching { beatmapsetRepository.save(beatmapset) }
        } else {
            Err(BeatmapsetNotFound)
        }
}
